src/traits/trait-singleton.php

```diff
@@ -0,0 +1,43 @@
+<?php
+/**
+ * Trait file for Singletons.
+ *
+ * @package wp-conditional-blocks
+ */
+
+declare( strict_types = 1 );
+
+namespace Alley\WP\WP_Conditional_Blocks\traits;
+
+/**
+ * Make a class into a singleton.
+ *
+ * @template T
+ */
+trait Singleton {
+	/**
+	 * Singleton instance for this class.
+	 *
+	 * @var self
+	 */
+	private static $instance = null;
+
+	/**
+	 * Get class instance.
+	 *
+	 * @return self
+	 */
+	public static function get_instance(): self {
+		if ( null === self::$instance ) {
+			self::$instance = new static();
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * Private constructor to prevent instantiation from outside the class.
+	 */
+	final private function __construct() {
+	}
```

Comment:
Not sure if we are planning to use this in other classes, but if those classes need to have a pseudo contructor to do initialization we could ass something like this:
```
public static function get_instance(): self {
    if ( null === self::$instance ) {
        self::$instance = new static();
        self::$instance->init();
    }

    return self::$instance;
}

/**
 * Use this function instead of constructor for any required initialization
 */
protected function init() {
    // Use in derived classes
}
```

Howver, if we aren't planning on using this in other classes we may not need this.

Reply:
I can add the init method. Everything I've read leads me to believe that I shouldn't leave the constructor public or undefined so as to prevent direct construction calls with the `new` operator. Let me know what you think before making a change.

Reply:
@rpasillas @attackant An init method in the `get_instance` is a good idea, but I don't think we should be adding it to the trait. Instead we should allow for it with something like

```php
<?php

trait Singleton {
    
    private static ?self $instance = null;
    
    public static function get_instance(): self {
		if ( null === self::$instance ) {
			self::$instance = new static();
			if ( method_exists( self::$instance, 'init' ) ) {
			    self::$instance->init();
			}
		}

		return self::$instance;
    }
}

class First {
    use Singleton;
}

class Second {
    use Singleton;
    
    protected function init(): void {
        echo 'This is the Second class.' . PHP_EOL;
    }
}

$first = First::get_instance();

$second = Second::get_instance();
```

Reply:
Thank you for the feedback @anubisthejackle. It came at the perfect time and I've included the fix in my latest PR for Issue #14.